kernel/dw1000.c

```diff
@@ -1435,6 +1443,91 @@ static void dw1000_timestamp(struct dw1000 *dw,
 #endif
 }
 
+/**
+ * dw1000_rx_link_qual() - Check RX quality and calculate KPIs
+ *
+ * @dw:			DW1000 device
+ */
+static void dw1000_rx_link_qual(struct dw1000 *dw)
+{
+	struct dw1000_rx *rx = &dw->rx;
+	uint32_t finfo, power, noise, ampl1, ampl2, ampl3;
+	uint32_t rxpacc, rxpsr, fpnrj, fppwr, overruns;
+	unsigned int snr, fpr, psr;
+
+	/* Extract measurements */
+	finfo = le32_to_cpu(rx->finfo);
+	ampl1 = le16_to_cpu(rx->time.fp_ampl1);
+	ampl2 = le16_to_cpu(rx->fqual.fp_ampl2);
+	ampl3 = le16_to_cpu(rx->fqual.fp_ampl3);
+	power = le16_to_cpu(rx->fqual.cir_pwr);
+	noise = le16_to_cpu(rx->fqual.std_noise);
+
+	/* Extra frame info */
+	rxpacc = DW1000_RX_FINFO_RXPACC(finfo);
+	rxpsr  = DW1000_RX_FINFO_RXPSR(finfo);
+
+	/* Get (expected) preamble symbol repetition count */
+	psr = dw1000_txpsrs[dw->txpsr];
+
+	/* Calculate RX signal estimates */
+	fpnrj = ampl1*ampl1 + ampl2*ampl2 + ampl3*ampl3;
+	fppwr = fpnrj / (rxpacc*rxpacc);
+
+	/* Calculate a simplistic S/N estimate */
+	snr = fppwr / (noise + 1);
+
+	/* Calculate a simplistic FP/Energy estimate */
+	fpr = (fpnrj / power) >> 9;
```

Comment:
Potential division by zero

Reply:
Also, there should probably be a #define for whatever the "9" represents

Reply:
yep